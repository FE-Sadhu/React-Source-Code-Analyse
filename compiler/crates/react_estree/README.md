# react_estree

This crate is a Rust representation of the [ESTree format](https://github.com/estree/estree/tree/master) and
popular extenions including JSX and (eventually) Flow and TypeScript.

This crate is intended as the main interchange format with outside code. A typical integration with React Compiler
will look as follows:

1. Host Compiler parses into the host AST format.
2. Host Compiler converts into `react_estree`.
3. Host Compiler invokes React Compiler to compile the input, which (conceptually) 
   returns the resulting code in `react_estree` format.
4. Host Compiler convert back from `react_estree` to its host AST format.

Because React Compiler is intended to support JavaScript-based toolchains, `react_estree` is designed to support
accurate serialization to/from estree-compatible JSON. We may also support the Babel AST format 
(a variant of ESTree) as well, depending on demand.