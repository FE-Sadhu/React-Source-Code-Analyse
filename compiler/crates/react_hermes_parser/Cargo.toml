[package]
name = "react_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
react_diagnostics = { workspace = true }
react_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
react_estree_codegen = { workspace = true }