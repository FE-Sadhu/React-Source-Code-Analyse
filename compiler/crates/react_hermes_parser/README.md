# react_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `react_estree` AST.