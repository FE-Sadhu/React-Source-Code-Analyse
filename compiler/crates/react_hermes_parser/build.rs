/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use react_estree_codegen::estree_hermes;

// Example custom build script.
fn main() {
    // Re-run if the codegen files change
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/codegen.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/lib.rs");
    println!("cargo:rerun-if-changed=../react_estree_codegen/src/ecmascript.json");
    println!("cargo:rerun-if-changed=../react_estree_codegen");

    let src = estree_hermes();
    let copyright = "
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
    "
    .to_string();
    let trimmed_copyright = copyright.trim();
    let contents = format!("{trimmed_copyright}\n{src}");
    std::fs::write("src/generated.rs", contents).unwrap();
}
