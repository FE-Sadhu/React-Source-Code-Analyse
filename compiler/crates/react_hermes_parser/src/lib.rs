/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

mod generated;
mod generated_extension;

use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;
use react_diagnostics::Diagnostic;
use react_estree::Program;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new(&buf);
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
