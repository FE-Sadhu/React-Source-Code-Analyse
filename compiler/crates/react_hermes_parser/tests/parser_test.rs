/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::env;

use insta::{assert_snapshot, glob};
use react_estree::SourceType;
use react_hermes_parser::parse;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let mut ast = parse(&input, path.to_str().unwrap()).unwrap();
        // TODO: hack to prevent changing lots of fixtures all at once
        ast.source_type = SourceType::Script;
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
