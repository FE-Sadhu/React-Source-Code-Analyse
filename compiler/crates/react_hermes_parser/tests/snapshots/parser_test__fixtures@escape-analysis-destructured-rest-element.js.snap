---
source: crates/react_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/react_hermes_parser/tests/fixtures/escape-analysis-destructured-rest-element.js
---
Input:
function Component(props) {
  // b is an object, must be memoized even though the input is not memoized
  const { a, ...b } = props.a;
  // d is an array, mut be memoized even though the input is not memoized
  const [c, ...d] = props.c;
  return <div b={b} d={d}></div>;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": [
          9,
          18
        ]
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": [
            19,
            24
          ]
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ObjectPattern",
                  "properties": [
                    {
                      "type": "Property",
                      "key": {
                        "type": "Identifier",
                        "name": "a",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          114,
                          115
                        ]
                      },
                      "value": {
                        "type": "Identifier",
                        "name": "a",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          114,
                          115
                        ]
                      },
                      "kind": "init",
                      "computed": false,
                      "shorthand": true,
                      "method": false,
                      "loc": null,
                      "range": [
                        114,
                        115
                      ]
                    },
                    {
                      "type": "RestElement",
                      "argument": {
                        "type": "Identifier",
                        "name": "b",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          120,
                          121
                        ]
                      },
                      "loc": null,
                      "range": [
                        117,
                        121
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    112,
                    123
                  ]
                },
                "init": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      126,
                      131
                    ]
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "a",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      132,
                      133
                    ]
                  },
                  "computed": false,
                  "loc": null,
                  "range": [
                    126,
                    133
                  ]
                },
                "loc": null,
                "range": [
                  112,
                  133
                ]
              }
            ],
            "loc": null,
            "range": [
              106,
              134
            ]
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "c",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": [
                        218,
                        219
                      ]
                    },
                    {
                      "type": "RestElement",
                      "argument": {
                        "type": "Identifier",
                        "name": "d",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          224,
                          225
                        ]
                      },
                      "loc": null,
                      "range": [
                        221,
                        225
                      ]
                    }
                  ],
                  "loc": null,
                  "range": [
                    217,
                    226
                  ]
                },
                "init": {
                  "type": "MemberExpression",
                  "object": {
                    "type": "Identifier",
                    "name": "props",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      229,
                      234
                    ]
                  },
                  "property": {
                    "type": "Identifier",
                    "name": "c",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": [
                      235,
                      236
                    ]
                  },
                  "computed": false,
                  "loc": null,
                  "range": [
                    229,
                    236
                  ]
                },
                "loc": null,
                "range": [
                  217,
                  236
                ]
              }
            ],
            "loc": null,
            "range": [
              211,
              237
            ]
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": [
                    248,
                    251
                  ]
                },
                "attributes": [
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "b",
                      "loc": null,
                      "range": [
                        252,
                        253
                      ]
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "b",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          255,
                          256
                        ]
                      },
                      "loc": null,
                      "range": [
                        254,
                        257
                      ]
                    },
                    "loc": null,
                    "range": [
                      252,
                      257
                    ]
                  },
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "d",
                      "loc": null,
                      "range": [
                        258,
                        259
                      ]
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "d",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": [
                          261,
                          262
                        ]
                      },
                      "loc": null,
                      "range": [
                        260,
                        263
                      ]
                    },
                    "loc": null,
                    "range": [
                      258,
                      263
                    ]
                  }
                ],
                "selfClosing": false,
                "loc": null,
                "range": [
                  247,
                  264
                ]
              },
              "children": [],
              "closingElement": {
                "type": "JSXClosingElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": [
                    266,
                    269
                  ]
                },
                "loc": null,
                "range": [
                  264,
                  270
                ]
              },
              "loc": null,
              "range": [
                247,
                270
              ]
            },
            "loc": null,
            "range": [
              240,
              271
            ]
          }
        ],
        "loc": null,
        "range": [
          26,
          273
        ]
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": [
        0,
        273
      ],
      "loc": null,
      "range": [
        0,
        273
      ]
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": [
    0,
    273
  ]
}
